src/Testing/TypeInferenceTestCase.php

```diff
@@ -107,78 +107,110 @@ public function assertFileAsserts(
+	}
+
 	/**
 	 * @api
 	 * @return array<string, mixed[]>
 	 */
 	public function gatherAssertTypes(string $file): array
 	{
-		$asserts = [];
-		$this->processFile($file, function (Node $node, Scope $scope) use (&$asserts, $file): void {
-			if (!$node instanceof Node\Expr\FuncCall) {
-				return;
-			}
+		$assertions = [];
 
-			$nameNode = $node->name;
-			if (!$nameNode instanceof Name) {
-				return;
-			}
-
-			$functionName = $nameNode->toString();
-			if ($functionName === 'PHPStan\\Testing\\assertType') {
-				$expectedType = $scope->getType($node->getArgs()[0]->value);
-				$actualType = $scope->getType($node->getArgs()[1]->value);
-				$assert = ['type', $file, $expectedType, $actualType, $node->getLine()];
-			} elseif ($functionName === 'PHPStan\\Testing\\assertNativeType') {
-				$nativeScope = $scope->doNotTreatPhpDocTypesAsCertain();
-				$expectedType = $nativeScope->getNativeType($node->getArgs()[0]->value);
-				$actualType = $nativeScope->getNativeType($node->getArgs()[1]->value);
-				$assert = ['type', $file, $expectedType, $actualType, $node->getLine()];
-			} elseif ($functionName === 'PHPStan\\Testing\\assertVariableCertainty') {
-				$certainty = $node->getArgs()[0]->value;
-				if (!$certainty instanceof StaticCall) {
-					$this->fail(sprintf('First argument of %s() must be TrinaryLogic call', $functionName));
-				}
-				if (!$certainty->class instanceof Node\Name) {
-					$this->fail(sprintf('ERROR: Invalid TrinaryLogic call.'));
-				}
-
-				if ($certainty->class->toString() !== 'PHPStan\\TrinaryLogic') {
-					$this->fail(sprintf('ERROR: Invalid TrinaryLogic call.'));
-				}
-
-				if (!$certainty->name instanceof Node\Identifier) {
-					$this->fail(sprintf('ERROR: Invalid TrinaryLogic call.'));
-				}
-
-				// @phpstan-ignore-next-line
-				$expectedertaintyValue = TrinaryLogic::{$certainty->name->toString()}();
-				$variable = $node->getArgs()[1]->value;
-				if (!$variable instanceof Node\Expr\Variable) {
-					$this->fail(sprintf('ERROR: Invalid assertVariableCertainty call.'));
-				}
-				if (!is_string($variable->name)) {
-					$this->fail(sprintf('ERROR: Invalid assertVariableCertainty call.'));
-				}
-
-				$actualCertaintyValue = $scope->hasVariableType($variable->name);
-				$assert = ['variableCertainty', $file, $expectedertaintyValue, $actualCertaintyValue, $variable->name];
-			} else {
-				return;
-			}
-
-			if (count($node->getArgs()) !== 2) {
-				$this->fail(sprintf(
-					'ERROR: Wrong %s() call on line %d.',
-					$functionName,
-					$node->getLine(),
-				));
+		foreach ($this->gatherAssertTypesLazy($file) as [$getAssertions]) {
+			foreach ($getAssertions() as $key => $assertion) {
+				$assertions[$key] = $assertion;
 			}
+		}
 
-			$asserts[$file . ':' . $node->getLine()] = $assert;
-		});
+		return $assertions;
+	}
 
-		return $asserts;
+	/**
+	 * @return array<string, array{callable(): iterable<string, mixed[]>}>
+	 */
+	public function gatherAssertTypesLazy(string $file): array
```

Comment:
Returns `[$file => callable(): iterable<string, mixed[]>]` so that we can use the method with yield from